zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1233 @@
+            let mut remaining_lazy_fees =
+                node.descendant_stake.bmul_bdiv(node.lazy_fees, node.total_stake()?)?;
+            // The fees that stay at this node.
+            let fees = node.lazy_fees.checked_sub_res(&remaining_lazy_fees)?;
+            self.mutate_node(index, |node| {
+                node.fees = node.fees.checked_add_res(&fees)?;
+                Ok(())
+            })?;
+            let (opt_lhs_index, opt_rhs_index) = self.children(index)?.into();
+            if let Some(lhs_index) = opt_lhs_index {
+                self.mutate_node(lhs_index, |lhs_node| {
+                    // The descendant's share of the stake:
+                    let child_lazy_fees = lhs_node
+                        .total_stake()?
+                        .bmul_bdiv(remaining_lazy_fees, node_descendant_stake)?;
+                    lhs_node.lazy_fees = lhs_node.lazy_fees.checked_add_res(&child_lazy_fees)?;
+                    remaining_lazy_fees = remaining_lazy_fees.checked_sub_res(&child_lazy_fees)?;
+                    Ok(())
+                })?;
+            }
+            if let Some(rhs_index) = opt_rhs_index {
+                self.mutate_node(rhs_index, |rhs_node| {
+                    rhs_node.lazy_fees =
+                        rhs_node.lazy_fees.checked_add_res(&remaining_lazy_fees)?;
+                    Ok(())
+                })?;
+            }
+        }
+        self.mutate_node(index, |node| {
+            node.lazy_fees = Zero::zero();
+            Ok(())
+        })?;
+        Ok(())
+    }
+
+    fn children(&self, index: u32) -> Result<LiquidityTreeChildIndices, DispatchError> {
+        let calculate_child =
+            |child_index: u32| Some(child_index).filter(|&i| i < self.node_count());
+        let left_child_index = index.checked_mul_res(&2)?.checked_add_res(&1)?;
+        let lhs = calculate_child(left_child_index);
+        let right_child_index = left_child_index.checked_add_res(&1)?;
+        let rhs = calculate_child(right_child_index);
+        Ok(LiquidityTreeChildIndices { lhs, rhs })
+    }
+
+    fn parent_index(&self, index: u32) -> Option<u32> {
+        if index == 0 { None } else { index.checked_sub(1)?.checked_div(2) }
+    }
+
+    fn path_to_node(&self, mut index: u32) -> Result<Vec<u32>, DispatchError> {
+        let mut path = Vec::new();
+        let mut iterations = 0;
+        let max_iterations = Self::max_depth().checked_add_res(&1)?;
+        while let Some(parent_index) = self.parent_index(index) {
+            if iterations == max_iterations {
+                return Err(LiquidityTreeError::MaxIterationsReached.into_dispatch_error::<T>());
+            }
+            path.push(index);
+            index = parent_index;
+            iterations = iterations.checked_add_res(&1)?;
+        }
+        path.push(0u32); // The tree's root is not considered in the loop above.
+        path.reverse(); // The path should be from root to the node
+        Ok(path)
+    }
```

Comment:
The function is defined recursively. Why was an iterative implementation chosen?

Recursive approach without depth check (as it is inherently included in `parent_index`):

```suggestion
    fn path_to_node(&self, index: u32) -> Result<Vec<u32>, DispatchError> {
        if index == 0 {
            Ok(vec![0u32])
        } else {
            let mut path = self.path_to_node(self.parent_index(index).unwrap_or(0u32))?;
            path.push(index);
            Ok(path)
        }
    }
```

Reply:
Arguably a defensive approach considering the depth is better though. In that case, the recursion would need another parameter `max_depth`.

Reply:
No change is requested by the way, this is just informative. If you like the suggested approach, feel free to apply it.

Reply:
Applied suggestion and then made the following changes:

1704bee2eee962282460d9527ab5eabaae8b3e06
52fa41093f48c807c132c3ba8f4499c2030d882a